\documentclass{article}
\usepackage[backend=bibtex, sorting=none]{biblatex}
\usepackage{hyperref}
\bibliography{references}
\begin{filecontents*}{references.bib}
@article{Tzelepis:2016ix,
author = {Tzelepis, Konstantinos and Koike-Yusa, Hiroko and De Braekeleer, Etienne and Li, Yilong and Metzakopian, Emmanouil and Dovey, Oliver M and Mupo, Annalisa and Grinkevich, Vera and Li, Meng and Mazan, Milena and Gozdecka, Malgorzata and Ohnishi, Shuhei and Cooper, Jonathan and Patel, Miten and McKerrell, Thomas and Chen, Bin and Domingues, Ana Filipa and Gallipoli, Paolo and Teichmann, Sarah and Ponstingl, Hannes and McDermott, Ultan and Saez-Rodriguez, Julio and Huntly, Brian J P and Iorio, Francesco and Pina, Cristina and Vassiliou, George S and Yusa, Kosuke},
title = {{A CRISPR Dropout Screen Identifies Genetic Vulnerabilities and Therapeutic Targets in Acute Myeloid Leukemia.}},
journal = {Cell reports},
year = {2016},
volume = {17},
number = {4},
pages = {1193--1205},
month = oct
}

@article{Iorio:2017,
author = {Iorio, Francesco and Behan, Fiona M and Goncalves, Emanuel and Beaver, Charlotte and Ansari, Rizwan  and Pooley, Rachel and Wilkinson, Piers and Harper, Sarah and Stronach, Euan and Saez-Rodriguez, Julio and Yusa, Kosuke and Garnett, Mathew J
},
title = {{Unsupervised correction of gene-independent cell responses to CRISPR-Cas9 targeting}},
journal = {revision},
volume = {0},
number = {0},
pages = {0--0},
month = Dec
}

@article{Li:2014kt,
author = {Li, Wei and Xu, Han and Xiao, Tengfei and Cong, Le and Love, Michael I and Zhang, Feng and Irizarry, Rafael A and Liu, Jun S and Brown, Myles and Liu, X Shirley},
title = {{MAGeCK enables robust identification of essential genes from genome-scale CRISPR/Cas9 knockout screens.}},
journal = {Genome Biology},
year = {2014},
volume = {15},
number = {12},
pages = {554}
}
\end{filecontents*}

\begin{document}
\title{CRISPRcleanR: An R package for unsupervised identification and correction of gene independent cell responses to CRISPR-cas9 targeting}
\author{Francesco Iorio, fi1@sanger.ac.uk}
\maketitle
\section{Quick start}

\subsection{Installation}

First, you need to install and load the devtools package. You can do this from CRAN. Invoke R and then type.

<<results='hide', eval=FALSE>>=
install.packages("devtools")
library(devtools)
@

Secondly, install the CRISPRcleanR with the following command:

<<results='hide', eval=FALSE>>=
install_github("francescojm/CRISPRcleanR")
@

\subsection{Raw sgRNA count median-ratio normalisation and computation of sgRNAs' log fold-changes}

Load the package.

<<results='hide'>>=
    library(CRISPRcleanR)
@

\textbf{Step 1:} Load your sgRNA library annotation. In this example we will use a built in data frame containing the annotation of the SANGER v1.0 library \cite{Tzelepis:2016ix}:
<<>>=
data(KY_Library_v1.0)
@

To use your own library annotation you will have to put it in a data frame with the same format of the \texttt{KY\char`_Library\char`_v1.0} data frame (detailed in the corresponding entry of the reference manual of the CRISPRcleanR package).\\

\textbf{Step 2:} Store the path of the tsv file containing your sgRNAs' raw counts in a temporare variable. In this example we will use counts generated upon a CRISPR-Cas9 pooled drop-out screen (described in \cite{Iorio:2017}) built in this package.
<<>>=
fn<-paste(system.file('extdata',package = 'CRISPRcleanR'),
          '/HT-29_counts.tsv',sep='')
@ 
The tsv with the sgRNAs' raw counts must be formatted as specified in the reference manual entry for function \texttt{ccr.NormfoldChanges}.\\

\textbf{Step 3:} Performing median-ratio normalisation of raw counts and computing sgRNAs' log fold-changes.
In this example we will exclude sgRNAs with less than 30 reads in the plasmid sample.
<<Norm_and_fc, fig.width=10, fig.height=5>>=
normANDfcs<-ccr.NormfoldChanges(fn,
                                min_reads=30,
                                EXPname='HT-29',
                                libraryAnnotation=KY_Library_v1.0)
@ 
This function returns a list of two data frames, respectively with normalised counts and log fold-changes, and it saves the as Robject in the directory whose path is specified with the parameter \texttt{outdir} (set to \texttt{'./'} by default).

<<size='tiny'>>=
head(normANDfcs$norm_counts)
head(normANDfcs$logFCs)
@ 

\textbf{IMPORTANT:} if there are control replicates in your sgRNAs count file their number must be specified by in the parameter \texttt{ncontrols} (equal to 1 by default) of the \texttt{ccr.NormfoldChanges} function.

\subsection{Genome sorting of sgRNAs' log fold-changes and their correction for gene independent responses to CRISPR-Cas9 targeting}

\textbf{Step 1:} Map genome-wide sgRNAs' log fold changes (averaged across replicates) on the genome, sorted according to their positions of the targeted region on the chromosomes.

<<>>=
gwSortedFCs<-
    ccr.logFCs2chromPos(normANDfcs$logFCs,KY_Library_v1.0)
@ 

<<size='tiny'>>=
head(gwSortedFCs)
@

\textbf{Step 2:} Identify and correct biased sgRNAs' log fold-changes putatively due to gene independent responses to CRISPR-Cas9 targeting (this function calls iteratively the \texttt{ccr.cleanChrm} function,
which performs the correction in each chromosome individually). In this example we are using a completely unsuerpvised approach and correcting chromosomal segments of equal sgRNA log fold-changes if they include sgRNAs targeting at least 3 different genes, and without making any assumption on gene essentiality nor knowing \textit{a priori} the copy number status of the included genes \cite{Iorio:2017}.

<<correction,fig.show='hide',  results='hide'>>=
correctedFCs<-ccr.GWclean(gwSortedFCs,display=TRUE,label='HT-29')
@
The corrected sgRNAs fold-changes are returned in a list (as a data frame), together with another data frame
with annotation of the identified segments and a vector of strings containing all the sgRNAs identifier genome-sorted.

<<size='tiny'>>=
    head(correctedFCs$corrected_logFCs)
@

Details on how the data frame with the corrected sgRNAs fold-changes should be interpreted can be found in the
entry of the \texttt{ccr.GWclean} function in the package reference manual.\\

This function also produces one plot per chromosome, with segments of sgRNAs' equal log fold-changes before and after the correction. An example of these plot is reported below (chromosome 8, in HT-29 with the region containing \textit{MYC} highly biased toward consistent negative fold-changes)

<<correction_chrm8, echo=FALSE, results='hide',fig.height=8,fig.width=6>>=
tmp<-ccr.cleanChrm(gwSortedFCs,8,display=TRUE,label='HT-29',
                        min.ngenes=3)
@

\subsection{Correcting sgRNAs' treatment counts for mean-variance modeling}
In order to apply the inverse transformation described in \cite{Iorio:2017}, thus to derive corrected normalised sgRNAs' treatment counts from corrected log fold-changes, it is sufficient to run the function \texttt{ccr.correctCounts} as follows:

<<results='hide'>>=
correctedCounts<-ccr.correctCounts('HT-29',
                                   normANDfcs$norm_counts,
                                   correctedFCs,
                                   KY_Library_v1.0,
                                   minTargetedGenes=3,
                                   OutDir='./')
@

With the plasmid counts, are suitable for mean-variance modeling approach (such that implemented in MAGeCK\cite{Li:2014kt}). 

<<>>=
head(correctedCounts)
@

This function also saves the correctedCounts as Rdata object at the location specified by the parameter \texttt{OutDir}.
To run MAGeCK, using these corrected sgRNAs' counts you will need to save them as a tsv file first:

<<>>=
write.table(correctedCounts,
            quote=FALSE,
            row.names = FALSE,
            sep='\t',
            file='./HT-29_mgk_input_corrected.tsv')
@

then use this file as input for MAGeCK.\\

\textbf{IMPORTANT:} the corrected sgRNAs' count are already median-normalised
therefore, when executing MAGeCK, the parameter \texttt{--norm-method} should be set to \texttt{none}.

\section{Visualisation and assessment of Results}
Coming Soon


\printbibliography

\end{document}