\documentclass{article}
\usepackage[backend=bibtex, sorting=none]{biblatex}
\usepackage{hyperref}
\bibliography{references}
\begin{filecontents*}{references.bib}
@article{Tzelepis:2016ix,
author = {Tzelepis, Konstantinos and Koike-Yusa, Hiroko and De Braekeleer, Etienne and Li, Yilong and Metzakopian, Emmanouil and Dovey, Oliver M and Mupo, Annalisa and Grinkevich, Vera and Li, Meng and Mazan, Milena and Gozdecka, Malgorzata and Ohnishi, Shuhei and Cooper, Jonathan and Patel, Miten and McKerrell, Thomas and Chen, Bin and Domingues, Ana Filipa and Gallipoli, Paolo and Teichmann, Sarah and Ponstingl, Hannes and McDermott, Ultan and Saez-Rodriguez, Julio and Huntly, Brian J P and Iorio, Francesco and Pina, Cristina and Vassiliou, George S and Yusa, Kosuke},
title = {{A CRISPR Dropout Screen Identifies Genetic Vulnerabilities and Therapeutic Targets in Acute Myeloid Leukemia.}},
journal = {Cell reports},
year = {2016},
volume = {17},
number = {4},
pages = {1193--1205},
month = oct
}

@article{Iorio:2017,
author = {Iorio, Francesco and Behan, Fiona M and Goncalves, Emanuel and Beaver, Charlotte and Ansari, Rizwan  and Pooley, Rachel and Wilkinson, Piers and Harper, Sarah and Stronach, Euan and Saez-Rodriguez, Julio and Yusa, Kosuke and Garnett, Mathew J
},
title = {{Unsupervised correction of gene-independent cell responses to CRISPR-Cas9 targeting}},
journal = {revision},
volume = {0},
number = {0},
pages = {0--0},
month = Dec
}

@article{Li:2014kt,
author = {Li, Wei and Xu, Han and Xiao, Tengfei and Cong, Le and Love, Michael I and Zhang, Feng and Irizarry, Rafael A and Liu, Jun S and Brown, Myles and Liu, X Shirley},
title = {{MAGeCK enables robust identification of essential genes from genome-scale CRISPR/Cas9 knockout screens.}},
journal = {Genome Biology},
year = {2014},
volume = {15},
number = {12},
pages = {554}
}

@article{Hart:2016ja,
author = {Hart, Traver and Moffat, Jason},
title = {{BAGEL: a computational framework for identifying essential genes from pooled library screens.}},
journal = {BMC bioinformatics},
year = {2016},
volume = {17},
pages = {164},
month = apr
}

@article{Subramanian:2005wu,
author = {Subramanian, A and Tamayo, P and Mootha, VK and Mukherjee, S and Ebert, BL and Gillette, MA and Paulovich, A and Pomeroy, SL and Golub, TR and Lander, ES},
title = {{Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles}},
journal = {Proceedings of the National Academy of Sciences of the United States of America},
year = {2005},
volume = {102},
number = {43},
pages = {15545}
}

@article{Iorio:2016ds,
author = {Iorio, Francesco and Knijnenburg, Theo A and Vis, Daniel J and Bignell, Graham R and Menden, Michael P and Schubert, Michael and Aben, Nanne and Gon{\c c}alves, Emanuel and Barthorpe, Syd and Lightfoot, Howard and Cokelaer, Thomas and Greninger, Patricia and van Dyk, Ewald and Chang, Han and de Silva, Heshani and Heyn, Holger and Deng, Xianming and Egan, Regina K and Liu, Qingsong and Mironenko, Tatiana and Mitropoulos, Xeni and Richardson, Laura and Wang, Jinhua and Zhang, Tinghu and Moran, Sebastian and Sayols, Sergi and Soleimani, Maryam and Tamborero, David and Lopez-Bigas, Nuria and Ross-Macdonald, Petra and Esteller, Manel and Gray, Nathanael S and Haber, Daniel A and Stratton, Michael R and Benes, Cyril H and Wessels, Lodewyk F A and Saez-Rodriguez, Julio and McDermott, Ultan and Garnett, Mathew J},
title = {{A Landscape of Pharmacogenomic Interactions in Cancer.}},
journal = {Cell},
year = {2016},
month = jul
}

\end{filecontents*}

\begin{document}
\title{CRISPRcleanR: An R package for unsupervised identification and correction of gene independent cell responses to CRISPR-cas9 targeting}
\author{Francesco Iorio, fi1@sanger.ac.uk}
\maketitle
\section{Quick start}

\subsection{Installation}

First, you need to install and load the devtools package. You can do this from CRAN. Invoke R and then type:

<<results='hide', eval=FALSE>>=
install.packages("devtools")
library(devtools)
@

Secondly, install CRISPRcleanR with the following command:

<<results='hide', eval=FALSE>>=
install_github("francescojm/CRISPRcleanR")
@

\subsection{Raw sgRNA count normalisation and computation of sgRNAs' log fold-changes}

Load CRISPRcleanR.
<<results='hide',echo=FALSE>>=
     options(warn = -1)
@

 <<results='hide'>>=
     library(CRISPRcleanR)
 @

<<results='hide',echo=FALSE>>=
     options(warn = 0)
@
%
\textbf{Step 1:} Load your sgRNA library annotation. In this example we will use a built in data frame containing the annotation of the SANGER v1.0 library, introduced in \cite{Tzelepis:2016ix}:
<<>>=
data(KY_Library_v1.0)
@
%
To use your own library annotation you will have to store it into a data frame with the same format of the \texttt{KY\char`_Library\char`_v1.0} (detailed in the corresponding entry of the reference manual of the CRISPRcleanR package).\\

\textbf{Step 2:} Store the path of the tsv file containing your sgRNAs' raw counts in a temporary variable. In this example we will use counts generated upon a CRISPR-Cas9 pooled drop-out screen (described in \cite{Iorio:2017}) built in this package, for an example immortalised human cancer cell line (HT-29)

<<>>=
 fn<-paste(system.file('extdata',package = 'CRISPRcleanR'),
           '/HT-29_counts.tsv',sep='')
@

The tsv file with the sgRNAs' raw counts must be formatted as specified in the reference manual entry of the \texttt{ccr.NormfoldChanges} function.\\

\textbf{Step 3:} Perform normalisation of raw counts and compute sgRNAs' log fold-changes (in this example we will exclude sgRNAs with less than 30 reads in the plasmid sample).

<<Norm_and_fc, fig.width=10, fig.height=5>>=
normANDfcs<-ccr.NormfoldChanges(fn,
                                 min_reads=30,
                                 EXPname='HT-29',
                                 libraryAnnotation=KY_Library_v1.0)
@

This function returns a list of two data frames containing, normalised counts and log fold-changes, respectively, and it saves them as Robjects in a user defined directory (specified by the parameter \texttt{outdir}, which is set to \texttt{'./'} by default).

<<size='small'>>=
head(normANDfcs$norm_counts)
head(normANDfcs$logFCs)
@

\textbf{IMPORTANT:} if there are control replicates in your sgRNAs count file their number must be specified by in the parameter \texttt{ncontrols} (equal to 1 by default) of the \texttt{ccr.NormfoldChanges} function.

\subsection{Genome sorting of sgRNAs' log fold-changes and their correction for gene independent responses to CRISPR-Cas9 targeting}

\textbf{Step 1:} Map genome-wide sgRNAs' log fold changes (averaged across replicates) on the genome, sorted according to the of their targeted gene on the chromosomes.

<<>>=
gwSortedFCs<-
     ccr.logFCs2chromPos(normANDfcs$logFCs,KY_Library_v1.0)
@

<<size='small'>>=
head(gwSortedFCs)
@
%
\textbf{Step 2:} Identify and correct biased sgRNAs' log fold-changes putatively due to gene independent responses to CRISPR-Cas9 targeting, using the \texttt{ccr.GWclean} function. This function calls iteratively the \texttt{ccr.cleanChrm} function, which performs the correction on an individual chromosome). In this example we are using a completely unsuerpvised approach and correcting chromosomal segments of equal sgRNA log fold-changes if they include sgRNAs targeting at least 3 different genes, and without making any assumption on gene essentiality, nor knowing \textit{a priori} the copy number status of the included genes \cite{Iorio:2017}.

<<correction,fig.show='hide',  results='hide'>>=
correctedFCs<-ccr.GWclean(gwSortedFCs,display=TRUE,label='HT-29')
@

The corrected sgRNAs fold-changes are returned in a list (as a data frame), together with the annotation of the identified segments (in another data frame) and a vector of strings containing all the gemome-sorted sgRNAs' identifiers.

<<size='small'>>=
     head(correctedFCs$corrected_logFCs)
@
%
Details on how the data frame with the corrected sgRNAs' log fold-changes should be interpreted can be found in the entry of the \texttt{ccr.GWclean} function, of the package reference manual.\\

This function also produces one plot per chromosome, with segments of sgRNAs' equal log fold-change before and after the correction. An example of these plot is reported below: chromosome 8, in HT-29, showing a region containing \textit{MYC}, which is highly biased toward consistent negative fold-changes.

<<correction_chrm8, echo=FALSE, results='hide',fig.height=8,fig.width=6>>=
tmp<-ccr.cleanChrm(gwSortedFCs,8,display=TRUE,label='HT-29',
                         min.ngenes=3)
@

\subsection{Correcting sgRNAs' treatment counts for mean-variance modeling}
In order to apply the inverse transformation described in \cite{Iorio:2017}, deriving corrected normalised sgRNAs' treatment counts from CRISPRcleanR corrected log fold-changes, it is sufficient to run the function \texttt{ccr.correctCounts} as follows:

<<results='hide'>>=
correctedCounts<-ccr.correctCounts('HT-29',
                                    normANDfcs$norm_counts,
                                    correctedFCs,
                                    KY_Library_v1.0,
                                    minTargetedGenes=3,
                                    OutDir='./')
@
%
Together with the plasmid counts, the corrected treatment counts can be used to compute depletion singificance scores with a mean-variance modeling approach (such that implemented in MAGeCK\cite{Li:2014kt}).

<<>>=
head(correctedCounts)
@

This function also saves the correctedCounts as Rdata object at the location specified by the parameter \texttt{OutDir}.
To run MAGeCK, using these corrected sgRNAs' counts you will need to save them as a tsv file, which will be ued as input for MAGeCK. This can be done by using the following function, which also returns the path of the saved file.

<<>>=
ccr.PlainTsvFile(correctedCounts,fprefix = 'HT-29')
@


\textbf{IMPORTANT:} the corrected sgRNAs' counts are already normalised,
therefore, when executing MAGeCK, the parameter \texttt{--norm-method} should be set to \texttt{none}.

\section{Visualisation and assessment of Results}

\subsection{Classification performances of reference sets of genes (or sgRNAs) based on depletion log fold-changes}

To perform a basic quality control assessment of your data it is possible to test the genome-wide profile of sgRNAs' depletion log fold-changes (logFCs) (or gene depletion logFCs, averaged across targeting sgRNAs) as a classifier of reference sets of core-fitness essential (CFE) and non-essential genes.\\

What you need for this is a named vector of sgRNAs (or gene) logFCs and two reference gene sets (respectively for positive and negative cases). In this example we make use of a precomputed essentiality profile from the builtin data object \texttt{EPLC.272HcorrectedFCs}. This is a list containing corrected sgRNAs log fold-changes and segment annotations for an example cell line (EPLC-272H), obtained using the \texttt{ccr.GWclean} function, as detailed in its reference manual entry. However the data frame containing the corrected log fold-changes, included in this list, reports also the original sgRNAs logFC (column \texttt{avgFC}), which will be used in this example.

<<>>=
data(EPLC.272HcorrectedFCs)
@

<<size='small'>>=
head(EPLC.272HcorrectedFCs$corrected_logFCs)
@

As reference gene sets we will use lists of CFE and non-essential genes assembled from multiple RNAi studies. These are used as classification template by the BAGEL algorithm to call gene depletion significance \cite{Hart:2016ja}, and are included in the builtin data objects \texttt{BAGEL\char`_essential} and \texttt{BAGEL\char`_nonEssential}.

<<>>=
data(BAGEL_essential)
data(BAGEL_nonEssential)

head(BAGEL_essential)
head(BAGEL_nonEssential)
@

Finally, you will need the sgRNAs library annotation. In this case we will use the builtin object \texttt{KY\char`_KY\char`_Library\char`_v1.0} (introduced in the previous section) \cite{Tzelepis:2016ix}. As for the previous examples, to use a different library annotation you will have to store it in a data frame with the same format of the \texttt{KY\char`_Library\char`_v1.0} data frame (detailed in the corresponding entry of the reference manual of the CRISPRcleanR package).\\

<<>>=
data(KY_Library_v1.0)
@

We will start with an evualuation at the sgRNA level. As mentioned, the logFCs need to be stored a named vector:

<<>>=
FCs<-EPLC.272HcorrectedFCs$corrected_logFCs$avgFC
names(FCs)<-rownames(EPLC.272HcorrectedFCs$corrected_logFCs)
@

The \texttt{ccr.genes2sgRNAs} function can be used, as follows, to convert the reference sets of CFE and non-essential genes into sets of sgRNAs:

<<results='hide',echo=FALSE>>=
    options(warn = -1)
@

<<>>=
BAGEL_essential_sgRNAs<-
    ccr.genes2sgRNAs(KY_Library_v1.0,BAGEL_essential)
BAGEL_nonEssential_sgRNAs<-
    ccr.genes2sgRNAs(KY_Library_v1.0,BAGEL_nonEssential)
@

<<results='hide',echo=FALSE>>=
    options(warn = 0)
@

Following these calls, possible warning messages could appear informing you that some of the reference genes are not targeted by any sgRNAs in the considered library. This has no impact on the following steps and results.\\

Finally, to visualise the ROC curve quantifying the performances in classifying the considered reference sets based on their logFCs, it is sufficient to call:
<<fig.width=5, fig.height=5>>=
sgRNA_level_ROC<-ccr.ROC_Curve(FCs,BAGEL_essential_sgRNAs,
                              BAGEL_nonEssential_sgRNAs)

@
%
To reperform the analysis at the gene level, you should first convert the profile of sgRNAs' logFCs into gene level summaries. The function \texttt{ccr.geneMeanFCs} performs this conversion by considering for each gene the average logFC across its targeting sgRNAs.

<<>>=
geneFCs<-ccr.geneMeanFCs(FCs,KY_Library_v1.0)
head(geneFCs)
@

The following call reperforms the ROC analysis at the gene level and it also computes and shows Recall values at fixed False Discovery Rate (which is defined by the user and in this case is equal to 5\%).

<<fig.width=5, fig.height=5>>=
gene_level_ROC<-ccr.ROC_Curve(geneFCs,
                         BAGEL_essential,
                         BAGEL_nonEssential,
                         FDRth = 0.05)
@


The same assessment can be performed at the level of Precision/Recall (PrRc) evaluation: to visualise the PrRc curve quantifying the performances in classifying the considered reference sets based on their logFCs, it is sufficient to call:
<<fig.width=5, fig.height=5>>=
sgRNA_level_PrRc<-ccr.PrRc_Curve(FCs,BAGEL_essential_sgRNAs,
                              BAGEL_nonEssential_sgRNAs)
@
%
As before, to reperform the analysis at the gene level, it is sufficient to call the following function, which also computes and shows Recall values at fixed False Discovery Rate (which is defined by the user and in this case is equal to 5\%).

<<fig.width=5, fig.height=5>>=
gene_level_PrRc<-ccr.PrRc_Curve(geneFCs,
                         BAGEL_essential,
                         BAGEL_nonEssential,
                         FDRth = 0.05)
@


As can be seen above, when setting the parameter \texttt{FRDth} to a value different from \texttt{NULL}
(its default value), this function also returns the log fold change threshold at which a classification FDR equal to the inputted value is achieved.

<<>>=
gene_level_PrRc$sigthreshold
@


\subsection{Depletion profile visualisation with genes signatures superimposed and recall computation}
For another quick assessment of your data, it is possible to visually inspect enrichments of predefined sets of core-fitness essential genes near the top of the genome wide essentiality profiles (composed of sgRNA or gene depletion logFCs ranked in increasing order), and to compute their classification recall at a fixed FDR (determined as deatiled in the previous subsection).\\

To this aim, in this example we will load additional sets of CFE genes assembled from MsigDB \cite{Subramanian:2005wu} as detailed in \cite{Iorio:2017}, and we will store all of them into a named list, as follows:

<<>>=
data(EssGenes.ribosomalProteins)
data(EssGenes.DNA_REPLICATION_cons)
data(EssGenes.KEGG_rna_polymerase)
data(EssGenes.PROTEASOME_cons)
data(EssGenes.SPLICEOSOME_cons)

SIGNATURES<-list(Ribosomal_Proteins=EssGenes.ribosomalProteins,
                 DNA_Replication = EssGenes.DNA_REPLICATION_cons,
                 RNA_polymerase = EssGenes.KEGG_rna_polymerase,
                 Proteasome = EssGenes.PROTEASOME_cons,
                 Spliceosome = EssGenes.SPLICEOSOME_cons,
                 CFE = BAGEL_essential,
                 non_essential = BAGEL_nonEssential)
@
%
Finally a visualisation of the gene essentiality profile with superimposed these signatures, can be created as follows:

<<fig.width=7, fig.height=7,fig.align='center'>>=
Recall_scores<-ccr.VisDepAndSig(FCsprofile = geneFCs,
                             SIGNATURES = SIGNATURES,
                             TITLE = 'EPLC-272H',
                             pIs = 6,
                             nIs = 7)
@
%
\textbf{IMPORTANT:} When calling \texttt{ccr.VisDepAndSig} it is important to correctly specify the index position of the reference gene sets that are used as classification template to derive the FDR threshold, within the list of signatures. In this case the template sets are \texttt{BAGEL\char`_essential} and \texttt{BAGEL\char`_nonEssential}, which in the \texttt{SIGNATURE} list are in position 6 and 7, respectively (this must be specified in the \texttt{pIs} and \texttt{nIs} parameters of the \texttt{ccr.VisDepAndSig} function).\\
%
This function also returns recall values at 5\% FDR for all the inputted signatures.

<<>>=
Recall_scores
@

\subsection{CRISPRcleanR correction assessment: Statistical tests}
To evaluate the effect of the CRISPRcleanR correction on your data it is possible to inspect the logFCs' variations for sgRNAs targeting predefined sets of genes for statistically significant differences with respect to background pre/post CRISPRcleanR correction.\\

To this aim, in this example we will use the builtin data object\\ \texttt{HT.29correctedFCs} containing corrected sgRNAs' logFCs and segment annotations for an example cell line (HT-29), obtained using the \texttt{ccr.GWclean} function, as detailed in its reference manual entry.

<<>>=
data(HT.29correctedFCs)
@

The function \texttt{ccr.perf\char`_statTests} performs this analysis, saving pdf figures in a user defined location ('./' by default).\\

Particularly, this functions assess statistical difference respect to background population pre/post CRISPRcleanR correction of logFCs for sgRNAs targeting respectively:

\begin{itemize}
        \item copy number (CN) deleted genes according to the GDSC1000 repository
        \item CN deleted genes (gistic score = -2) according to the CCLE repository
        \item non expressed genes (FPKM lower than 0.05)
        \item genes with gistic score = 1
        \item genes with gistic score = 2
        \item non espressed genes (FPKM lower than 0.05) with gistic score = 1
        \item non espressed genes (FPKM lower than 0.05) with gistic score = 2
        \item genes with minimal CN = 2, according to the GDSC1000
        \item genes with minimal CN = 4, according to the GDSC1000
        \item genes with minimal CN = 8, according to the GDSC1000
        \item genes with minimal CN = 10, according to the GDSC1000
        \item non expressed genes (FPKM lower than 0.05) with minimal CN = 2, according to the GDSC1000
        \item non expressed genes (FPKM lower than 0.05) with minimal CN = 4, according to the GDSC1000
        \item non expressed genes (FPKM lower than 0.05) with minimal CN = 8, according to the GDSC1000
        \item non expressed genes (FPKM lower than 0.05) with minimal CN = 10, according to the GDSC1000

        \end{itemize}

The call should be as follows:

<<>>=
RES<-ccr.perf_statTests('HT-29',libraryAnnotation = KY_Library_v1.0,
                   correctedFCs = HT.29correctedFCs$corrected_logFCs,
                   GDSC.geneLevCNA = NULL,
                   CCLE.gisticCNA = NULL,
                   RNAseq.fpkms = NULL)
@

This saves a pdf with the following figure in a user defined location ('./' by default).

\begin{figure}[!h]
\centering
\includegraphics[width=130mm]{HT-29_bp.pdf}
\end{figure}

Leaving the parameters \texttt{GDSC.geneLevCNA}, \texttt{CCLE.gisticCNA}, and\\ \texttt{RNAseq.fpkms} to their default NULL value will force this function to use the respective builtin data objects encompassing data only for 15 cell lines, used in \cite{Iorio:2017} and in this package to assess the performances of CRISPRcleanR.\\

\textbf{IMPORTANT:} To analyse data from screening a different cell line ad-hoc \texttt{GDSC.geneLevCNA}, \texttt{CCLE.gisticCNA}, and \texttt{RNAseq.fpkms} data objects should be assembled and passed to this function. These should have the same format of the respective builtin counterparts, detailaed in their user reference manual entries, which contains also additional infos on how to derive this datasets for 1,000 human cance cell lines from the GDSC1000 data portal \cite{Iorio:2016ds}).\\

Comprehensive statistical scores (detailed in the user reference manual) are also returned in output by this function.\\

Another example, analysing in the same way the essentiality profile of the EPLC-272H cell line (includede in a corresponding builtin R object) is reported below,

<<>>=
RES<-ccr.perf_statTests('EPLC-272H',libraryAnnotation = KY_Library_v1.0,
                    correctedFCs = EPLC.272HcorrectedFCs$corrected_logFCs)
@

<<size='small'>>=
RES$PVALS
RES$EFFsizes
@

\begin{figure}[!h]
\centering
\includegraphics[width=130mm]{EPLC-272H_bp.pdf}
\end{figure}

\newpage
To inspect the variation induced by the CRISPRcleanR correction on the logFCs' distributions of sgRNAs targeting defined sets of genes prior/post\\ CRISPRcleanR correction, the following function can be also used (and will produce the following density plots):

<<>>=
ccr.perf_distributions('HT-29',HT.29correctedFCs$corrected_logFCs,
                       libraryAnnotation = KY_Library_v1.0)
@

\textbf{IMPORTANT:} The instructions provided regarding what CN/transcriptional data object to pass to the \texttt{ccr.perf\char`_statTests} apply also to this function.\\

Additional infos on how to use this function can be found in the user reference manual.

\subsection{Recall variations following CRISPRcleanR correction for reference, copy number amplified, and non expressed genes}

A final analysis that can be done with the CRISPRcleanR package in order to evalute the effect of its correction on the classfication recall for predefined gene sets can be performed by calling the function, which can work at the sgRNA

<<>>=
ccr.RecallCurves('EPLC-272H',EPLC.272HcorrectedFCs$corrected_logFCs,
                  libraryAnnotation=KY_Library_v1.0)
@

as well as the gene level:

<<>>=
ccr.RecallCurves('EPLC-272H',EPLC.272HcorrectedFCs$corrected_logFCs,
                  libraryAnnotation=KY_Library_v1.0,GeneLev = TRUE)
@

\textbf{IMPORTANT:} The instructions provided regarding what CN/transcriptional data object to pass to the \texttt{ccr.perf\char`_statTests} apply also to this function.\\

Additional infos on how to use this function can be found in the user reference manual.

\subsection{CRISPRcleanR correction assessment: impact on phenotype and possible distortion}
To evaluate the effect of the CRISPRcleanR correction on the genes showing a significant loss/gain-of-fitness
effect (fitness genes) in the uncorrected data, a comparison of fitness gene sets (computed with MAGeCK \cite{Li:2014kt}) before/after CRISPRcleanR correction can be performed as follows\\

Before running the below example let us clean the R enviromnent with the following command:
<<>>=
rm(list=ls())
@

Subsequently, as detailed above, we load the reference sgRNA library annotation, we load and normalise sgRNA counts for the EPLC-272H example cell line.

<<fig.show='hide',  results='hide'>>=
data(KY_Library_v1.0)
  
fn<-paste(system.file('extdata', package = 'CRISPRcleanR'),
          '/EPLC-272H_counts.tsv',sep='')

normANDfcs<-ccr.NormfoldChanges(fn,min_reads=30,
                                EXPname='EPLC-272H',
                                libraryAnnotation = KY_Library_v1.0)
@

Then, we save the normalised counts as plain tsv file (suitable for MAGeCK):

<<>>=
uncorrected_fn<-
    ccr.PlainTsvFile(sgRNA_count_object = normANDfcs$norm_counts,
                     fprefix = 'EPLC-272H')
@

At this point we execute MAGeCK on the uncorrected normalised counts. \textbf{IMPORTANT:} This requires python and MAGeCK (v0.5.3). Additionally, as these counts are already normalised, we should specify that the normalisation shouldn't be reperformed by MAGeCK.\\

Run-time information displayed by MAGeCK (not reported in this document) will be visualised and all the outputted files will be saved in the working directory (by default), unless different specified in the corresponding argument (see reference manual entry for this function for further details).

<<results='hide'>>=
uncorrected_gs_fn<-
    ccr.ExecuteMageck(mgckInputFile = uncorrected_fn,
                      expName = 'EPLC-272H',
                      normMethod = 'none')
@

This function returns in output the path of the gene summary file outputted by MAGeCK.

<<>>=
uncorrected_gs_fn
@

As detailed in the previous sections of this document, we now correct the sgRNA counts of the example cell lines with CRISPRcleanR and save them in a plain tsv file (suitable to be used by MAGeCK)

<<fig.show='hide',  results='hide'>>=
gwSortedFCs<-ccr.logFCs2chromPos(normANDfcs$logFCs,KY_Library_v1.0)

correctedFCs<-ccr.GWclean(gwSortedFCs,display=FALSE,label='EPLC-272H')

correctedCounts<-ccr.correctCounts('EPLC-272H',normANDfcs$norm_counts,
                                   correctedFCs,
                                   KY_Library_v1.0,
                                   minTargetedGenes=3,
                                   OutDir='./')

corrected_fn<-ccr.PlainTsvFile(sgRNA_count_object = correctedCounts,
                               fprefix = 'EPLC-272H_ccleaned')
@

Then we execute MAGeCK on the corrected sgRNA counts.

<<results='hide'>>=
corrected_gs_fn<-ccr.ExecuteMageck(mgckInputFile = corrected_fn,
                                   expName = 'EPLC-272H_ccleaned',
                                   normMethod = 'none')
@

Finally, we assess the impact of CRISPcleanR correction on gain/loss-of-fitness genes with the following call, which
will produce bar/pie charts reported below
<<fig.width=7, fig.height=7,fig.align='center'>>=
RES<-ccr.impactOnPhenotype(MO_uncorrectedFile = uncorrected_gs_fn,
                      MO_correctedFile = corrected_gs_fn,
                      expName = 'EPLC-272H')
@

\printbibliography

\end{document}